app_check/integration_test/src/integration_test.cc

```diff
@@ -821,4 +919,191 @@ TEST_F(FirebaseAppCheckTest, TestFunctionsFailure) {
+          [&](const firebase::firestore::DocumentSnapshot& result,
+              firebase::firestore::Error error_code,
+              const std::string& error_message) {
+            EXPECT_EQ(error_code, firebase::firestore::kErrorOk);
+            EXPECT_EQ(error_message, "");
+            document_snapshots.push_back(result.GetData());
+          });
+
+  WaitForCompletion(
+      document.Set(firebase::firestore::MapFieldValue{
+          {"val", firebase::firestore::FieldValue::String("update")}}),
+      "document.Set 1");
+
+  registration.Remove();
+  WaitForCompletion(
+      document.Set(firebase::firestore::MapFieldValue{
+          {"val", firebase::firestore::FieldValue::String("final")}}),
+      "document.Set 2");
+  EXPECT_THAT(
+      document_snapshots,
+      testing::ElementsAre(
+          firebase::firestore::MapFieldValue{
+              {"val", firebase::firestore::FieldValue::String("start")}},
+          firebase::firestore::MapFieldValue{
+              {"val", firebase::firestore::FieldValue::String("update")}}));
+}
+
+TEST_F(FirebaseAppCheckTest, TestFirestoreTransactionFailure) {
+  // Don't set up AppCheck
+  InitializeApp();
+  InitializeFirestore();
+
+  firebase::firestore::DocumentReference document = CreateFirestoreDoc();
+
+  bool received_pending_cache = false;
+  bool received_permission_denied = false;
+  firebase::firestore::ListenerRegistration registration =
+      document.AddSnapshotListener(
+          [&received_pending_cache, &received_permission_denied](
+              const firebase::firestore::DocumentSnapshot& result,
+              firebase::firestore::Error error_code,
+              const std::string& error_message) {
+            // We first expect one call for a successful change to the cache.
+            if (error_code == firebase::firestore::kErrorNone) {
+              EXPECT_FALSE(received_pending_cache);
+              EXPECT_TRUE(result.metadata().has_pending_writes());
+              EXPECT_TRUE(result.metadata().is_from_cache());
+              received_pending_cache = true;
+            } else {
+              // We expect one call with a Permission Denied error, from the
+              // server.
+              EXPECT_FALSE(received_permission_denied);
+              EXPECT_EQ(error_code,
+                        firebase::firestore::kErrorPermissionDenied);
+              received_permission_denied = true;
+            }
+          });
+
+  WaitForCompletion(
+      document.Set(firebase::firestore::MapFieldValue{
+          {"val", firebase::firestore::FieldValue::String("transaction")}}),
+      "document.Set transaction", firebase::firestore::kErrorPermissionDenied);
+
+  registration.Remove();
+
+  WaitForCompletion(
+      document.Set(firebase::firestore::MapFieldValue{
+          {"val", firebase::firestore::FieldValue::String("final")}}),
+      "document.Set final", firebase::firestore::kErrorPermissionDenied);
+
+  EXPECT_TRUE(received_pending_cache);
```

Comment:
I think this test will be flaky. I can't see that there is any guarantee that the permission error would have occurred by this point, especially in the case of a slow network. Instead, I'd recommend to use a condition variable and notify it once the permission denied error is received by the snapshot listener, timing out after some reasonable amount of time (like 30 seconds).

Reply:
Done